[REVIEW] CSV Reader: Use type dispatcher instead of switch block

j-ieong opened the pull request:
This is an enhancement to use the type dispatcher, as suggested by @jrhemstad.

* Refactor data conversion calls to use template functions with dispatcher
* Re-use parsing opts struct to pass parameters to conversion funcs
* Remove multiple usage of parsing opts struct and instead assign values once
* Replace Simple gtest with mixed integer and float test

<!--

Thanks for wanting to contribute to cuDF :)

First, if you need some help or want to chat to the core developers, please
visit https://rapids.ai/community.html for links to our Google Group and other
communication channels.

Here's some guidelines to help the review process go smoothly.

0. Please write a description in this text box of the changes that are being
   made.

1. Please ensure that you have written units tests for the changes made/features
   added.

2. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

3. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]`.

4. If your pull request is ready to be reviewed without requiring additional
   work on top of it, then remove the `[WIP]` label (if present) and replace
   it with `[REVIEW]`. If assistance is required to complete the functionality,
   for example when the C/C++ code of a feature is complete but Python bindings
   are still required, then add the label `[HELP-REQ]` so that others can triage
   and assist. The additional changes then can be implemented on top of the
   same PR. If the assistance is done by members of the rapidsAI team, then no
   additional actions are required by the creator of the original PR for this,
   otherwise the original author of the PR needs to give permission to the
   person(s) assisting to commit to their personal fork of the project. If that
   doesn't happen then a new PR based on the code of the original PR can be
   opened by the person assisting, which then will be the PR that will be
   merged.

5. Once all work has been done and review has taken place please do not add
   features or make changes out of the scope of those requested by the reviewer
   (doing this just add delays as already reviewed code ends up having to be
   re-reviewed/it is hard to tell what is new etc!). Further, please do not
   rebase your branch on master/force push/rewrite history, doing any of these
   causes the context of any comments made by reviewers to be lost. If
   conflicts occur against master they should be resolved by merging master
   into the branch used for making the pull request.

Many thanks in advance for your cooperation!

-->

jrhemstad commented on cpp/src/io/csv/type_conversion.cuh:
```diff
@@ -332,8 +200,94 @@ int32_t convertStrtoHash(const char * key, long start_idx, long end_idx, uint32_
+__host__ __device__ gdf_date32 parseDateFormat(char *data, long start_idx, long end_idx, bool dayfirst);
+__host__ __device__ gdf_date64 parseDateTimeFormat(char *data, long start_idx, long end_idx, bool dayfirst);
+
+typedef struct parsing_opts_ {
```
Does this struct need to be a C struct for CFFI reasons? If not, we should make it a C++ struct. You can then provide in-struct default values for each member.

j-ieong replied:
Nope, this is an internal struct so it can be C++. I've changed how the CSV reader uses this struct (populating it from read_csv_args once now) so we don't need the default values as we assign each value.

jrhemstad commented on cpp/src/io/csv/csv_reader.cu:
```diff
@@ -1067,85 +1064,25 @@ __global__ void convertCsvToGdf(
+					stringCol++;
+				}
+				else {
+					cudf::type_dispatcher(
```
Love it. Hopefully we'll get strings figured out soon so that we can also dispatch for `GDF_STRING`. In the mean time, your workaround is a good one.

jrhemstad commented on cpp/src/io/csv/type_conversion.cuh:
```diff
@@ -332,8 +200,94 @@ int32_t convertStrtoHash(const char * key, long start_idx, long end_idx, uint32_
+
+template<>
+__host__ __device__
+cudf::date32 convertStrToValue<cudf::date32>(const char *data, long start, long end, const parsing_opts_t& opts) {
```
Love this! Happy to see that being able to specialize for the non-fundamental types is being used.

jrhemstad commented on cpp/src/io/csv/type_conversion.cuh:
```diff
@@ -332,8 +200,94 @@ int32_t convertStrtoHash(const char * key, long start_idx, long end_idx, uint32_
+__host__ __device__
+cudf::date32 convertStrToValue<cudf::date32>(const char *data, long start, long end, const parsing_opts_t& opts) {
+
+	return cudf::date32{ parseDateFormat(const_cast<char*>(data), start, end, opts.dayfirst) };
```
Would it be more appropriate to make the `data` parameter for `parseDateFormat` to be `const` rather than having to use `const_cast`. 

Based on a cursory examination, it does not appear that `parseDateFormat` should be able to modify `data`, and therefore for `const` correctness, should be marked as `const`.

j-ieong replied:
Changed the parameters to be const char* in the child functions. Thanks!

jrhemstad commented on cpp/src/io/csv/type_conversion.cuh:
```diff
@@ -332,8 +200,94 @@ int32_t convertStrtoHash(const char * key, long start_idx, long end_idx, uint32_
+__host__ __device__
+cudf::date64 convertStrToValue<cudf::date64>(const char *data, long start, long end, const parsing_opts_t& opts) {
+
+	return cudf::date64{ parseDateTimeFormat(const_cast<char*>(data), start, end, opts.dayfirst) };
```
`parseDateTimeFormat` should be updated such that the `data` parameter is `const` for `const` correctness.

j-ieong replied:
Changed it to be const char*, thanks!

jrhemstad commented on cpp/src/io/csv/csv_reader.cu:
```diff
@@ -53,12 +53,9 @@
-constexpr int32_t HASH_SEED = 33;
+#include "rmm/rmm.h"
 
 using namespace std;
```
Avoid using the `using namespace` at global scope. This almost always ends up leading to hard to track down problems.

j-ieong replied:
Looks like someone else has already removed it in the latest code.

jrhemstad commented on cpp/src/io/csv/csv_reader.cu:
```diff
@@ -960,14 +948,15 @@ gdf_error launch_dataConvertColumns(raw_csv_t *raw_csv, void **gdf, gdf_valid_ty
-	opts.thousands		= raw_csv->thousands;
-
-	convertCsvToGdf <<< gridSize, blockSize >>>(
+	opts.delimiter			= raw_csv->delimiter;
```
If `parsing_opts_t` were a C++ struct, there could be a `parsing_opts_t` constructor that accepts a `raw_csv_t` that does this initialization.

j-ieong replied:
I've changed the usage of this struct so we only use it once now in read_csv(). I can still make a constructor, but I didnt't want to move the assignment code into a different file for now.

j-ieong commented on cpp/src/io/csv/type_conversion.cuh:
```diff
@@ -332,8 +200,94 @@ int32_t convertStrtoHash(const char * key, long start_idx, long end_idx, uint32_
+__host__ __device__ gdf_date32 parseDateFormat(char *data, long start_idx, long end_idx, bool dayfirst);
+__host__ __device__ gdf_date64 parseDateTimeFormat(char *data, long start_idx, long end_idx, bool dayfirst);
+
+typedef struct parsing_opts_ {
```
Done!

j-ieong commented on cpp/src/io/csv/csv_reader.cu:
```diff
@@ -138,7 +126,7 @@ gdf_error launch_dataTypeDetection(raw_csv_t * raw_csv, long row_offset, column_
 __global__ void countRecords(char *data, const char terminator, const char quotechar, long num_bytes, long num_bits, unsigned long long* num_records);
 __global__ void storeRecordStart(char *data, const char terminator, const char quotechar, long num_bytes, long num_bits, unsigned long long* num_records,unsigned long long* recStart) ;
-__global__ void convertCsvToGdf(char *csv, const parsing_opts_t opts, unsigned long long num_records, int num_columns,bool *parseCol,unsigned long long *recStart,gdf_dtype *dtype,void **gdf_data,gdf_valid_type **valid,string_pair **str_cols,unsigned long long row_offset, long header_row,bool dayfirst,unsigned long long *num_valid);
+__global__ void convertCsvToGdf(char *csv, const parsing_opts_t opts, unsigned long long num_records, int num_columns,bool *parseCol,unsigned long long *recStart,gdf_dtype *dtype,void **gdf_data,gdf_valid_type **valid,string_pair **str_cols,unsigned long long row_offset, long header_row,unsigned long long *num_valid);
```
Changed. Thanks!

j-ieong commented:
@jrhemstad @harrism The type dispatcher is certainly nice and intuitive to use. I've updated the PR to add change another switch block to use the type dispatcher for getting size of data type during memory allocation. I also clean-up a bit more things and added more comments.

I changed the parsing options struct to be assigned once, rather than for each kernel launch and having the parsing options be duplicated in both raw_csv and parsing_opts. I've create a constructor for now, since I didn't want to move the assignment code to another file just yet.

j-ieong commented:
rerun tests

vuule commented on cpp/src/io/csv/csv_reader.cu:
```diff
@@ -919,60 +876,40 @@ gdf_error uploadDataToDevice(const char* h_uncomp_data, size_t h_uncomp_size, ra
+ *---------------------------------------------------------------------------**/
+struct SizeOfFunctor {
+	template <typename T>
+	int operator()() {
```
You can make this constexpr to allow use in any context where sizeof can be used.
Edit: might need a constexpr constructor to really work in constexpr context.

j-ieong replied:
Sure, thanks!

jrhemstad commented:
> Quality PR.
> One question: do you know if the type dispatcher supports lambdas?

@vuule 

The way the `type_dispatcher` works is it expects a callable object with an `operator()` template.

The only way to generate a lambda with an `operator()` template is via C++14 generic lambdas. 

For example:
```
auto lambda = [] (auto a) { return a; };
```

is effectively the same as:

```
class /* unnamed */
{
public:
    template<typename T>
    T operator () (T a) const { return a; }
};
```

So you _could_ do something like...
```
auto lambda = [] (auto dummy, gdf_column * col) { 
   using ColumnType = decltype(dummy);
   ColumnType * typed_data = static_cast<ColumnType*>(col->data);
   ....
 };
```
But I really do not suggest you do this. It's much better to just use a named functor `struct`.

jrhemstad commented on cpp/src/io/csv/datetime_parser.cuh:
```diff
@@ -64,15 +64,29 @@
+// Simplified to handle only numeric characters and returns positive numbers
+template<typename T>
+__host__ __device__
+T convertStrToInteger(const char *data, long start, long end) {
```
Update documentation to use the standard Doxygen format from our template: https://github.com/rapidsai/cudf/blob/branch-0.6/cpp/src/examples/documentation_example.cpp

jrhemstad commented on cpp/src/io/csv/csv_reader.cu:
```diff
@@ -1259,33 +1191,77 @@ gdf_error launch_dataConvertColumns(raw_csv_t *raw_csv, void **gdf, gdf_valid_ty
+	template <typename T,
+	          typename std::enable_if_t<!std::is_integral<T>::value>* = nullptr>
+	__host__ __device__ __forceinline__
+	void operator()(const char* csvData, void *gdfColumnData, long rowIndex,
```
Adding Doxygen documentation to both template specializations would help in understanding what's going on here.

jrhemstad commented on cpp/src/io/csv/csv_reader.cu:
```diff
@@ -285,17 +238,21 @@ gdf_error read_csv(csv_read_arg *args)
 		}
 		raw_csv->d_falseValues = h_values;
 	}
+	raw_csv->opts.trueValues       = thrust::raw_pointer_cast(raw_csv->d_trueValues.data());
```
I think this is much cleaner.
```suggestion
	raw_csv->opts.trueValues       = raw_csv->d_trueValues.data().get();
```

jrhemstad commented on cpp/src/io/csv/csv_reader.cu:
```diff
@@ -285,17 +238,21 @@ gdf_error read_csv(csv_read_arg *args)
 	}
+	raw_csv->opts.trueValues       = thrust::raw_pointer_cast(raw_csv->d_trueValues.data());
+	raw_csv->opts.trueValuesCount  = raw_csv->d_trueValues.size();
+	raw_csv->opts.falseValues      = thrust::raw_pointer_cast(raw_csv->d_falseValues.data());
```
```suggestion
	raw_csv->opts.falseValues      = raw_csv->d_falseValues.data().get();
```

BradReesWork commented:
rerun tests

j-ieong commented:
rerun tests
